i2c lockup fix

9names opened the pull request:
When doing an i2c WriteRead the driver would lock up indefinitely due to an error in how the transmit FIFO was being checked.
I also removed asserts for zero length read and writes, since it's quite legal for someone to do a zero length write to check for the existence of a peripheral on the bus, and some devices do not have any registers, so they do not require writes in WriteRead.
Also improves error handling.
Fixes #95

9names commented on rp2040-hal/src/i2c.rs:
```diff
@@ -309,7 +309,7 @@ macro_rules! hal {
                         let last = i == bytes.len() - 1;
 
-                        while 16 - self.i2c.ic_txflr.read().txflr().bits() > 0 {}
+                        while self.i2c.ic_txflr.read().txflr().bits() > 0 {}
```
Yep that makes sense, I managed to completely misread the intent here.
It's enough magic that it deserves a comment, I'll add one of those too.

9names replied:
How would you feel about helper functions?
```rust
const TX_FIFO_SIZE: u8 = 16;

impl<PINS> I2C<$I2CX, PINS> {
    // Number of bytes currently in the TX FIFO
    #[inline]
    fn tx_fifo_used(&self) -> u8 {
        self.i2c.ic_txflr.read().txflr().bits() 
    }

    // Remaining capacity in the TX FIFO
    #[inline]
    fn tx_fifo_free(&self) -> u8 {
        self.tx_fifo_used() - TX_FIFO_SIZE
    }

    // TX FIFO is at capacity
    #[inline]
    fn tx_fifo_full(&self) -> bool {
        self.tx_fifo_free() != 0
    }
}
```

thejpster replied:
Bonza. I tried, but failed to insert them in the correct place within the macro soup to get them to compile.

thejpster replied:
But triple slashes please :)

anall commented:
>I also removed asserts for zero length read and writes, since it's quite legal for someone to do a zero length write

I can't find it in the datasheet now when looking, but I know I read that the RP2040 does not allow doing a zero-length I2C write (with the R/W bit of the address set to 0) -- when doing a write, the contents of `dat` will always be written after the address (if a peripheral ACKs)

This has been tested empirically -- doing an I2C transfer with `STOP` set, `CMD` set to 0 (write), and `DAT` containing whatever: sends the address, gets the ACK, sends `DAT`, and then sends stop.

A "zero-length" write, and then a read would also be impossible unless we can skip/elide the write.

I'm not actually sure if a zero-length read is possible, and haven't tested it.

thejpster commented:
Ahh, if you do write_read with a zero length write you are in fact doing a read, and so the R/W bit will be one.

I'm happy to have `write`, `read` and `write_read` methods which take non zero length buffers - it's clear what they are for.

anall commented:
> GIven that, how does their bus scan example work?

Bitbang I2C

thejpster commented:
> > GIven that, how does their bus scan example work?
> 
> Bitbang I2C

I'm not sure that's right...

https://github.com/raspberrypi/pico-examples/blob/master/i2c/bus_scan/bus_scan.c#L69

anall commented:
> I'm not sure that's right...
> 
> https://github.com/raspberrypi/pico-examples/blob/master/i2c/bus_scan/bus_scan.c#L69

I know something else referenced doing bitbang I2C. but if you look at line 58 of that file:

        // Perform a 1-byte dummy read from the probe address. If a slave
        // acknowledges this address, the function returns the number of bytes
        // transferred. If the address byte is ignored, the function returns
        // -1.

9names commented:
Since this didn't get merged quickly I added the missing Read function, and replaced the asserts on Read/Write/WriteRead with Error types.
Read worked well enough for a simple i2c-detect function, and still worked with my mag3110 magnetometer.
```rust
let sda_pin = pins.gpio18.into_mode::<FunctionI2C>();
let scl_pin = pins.gpio19.into_mode::<FunctionI2C>();

let mut i2c = I2C::i2c1(
    pac.I2C1,
    sda_pin,
    scl_pin,
    400.kHz(),
    &mut pac.RESETS,
    SYS_HZ.Hz(),
);

for i in 0..255 {
    let mut readbuf: [u8; 1] = [0; 1];
    let result = i2c.read(i, &mut readbuf);
    if let Ok(d) = result {
        info!("Device found at address{:?}", i)
    }
}
```

thejpster commented on rp2040-hal/src/i2c.rs:
```diff
@@ -23,6 +23,14 @@ use rp2040_pac::{I2C0, I2C1, RESETS};
+    /// Target i2c address is out of range
+    AddressOutOfRange(u32),
+    /// Target i2c address is reserved
+    AddressReserved(u32),
```
u32 should be u8 here, as that's the size we use for an I2C address.

9names replied:
Yep, true. I was quite tired when I implemented this, thought I should keep the types the same for some reason.
I've updated it to be u8 now.

thejpster commented on rp2040-hal/src/i2c.rs:
```diff
@@ -23,6 +23,14 @@ use rp2040_pac::{I2C0, I2C1, RESETS};
     /// I2C abort with error
     Abort(u32),
+    /// User passed in a read buffer that was 0 or >255 length
+    InvalidReadBufferLength(u32),
```
u32 should be usize here, as that's the type for a slice length.

9names replied:
The buffer length enum variants are now usize too.

thejpster reviewed (changes requested):
Just need to clean up the error type, then this is good to go.

thejpster commented:
Excited to try the read function (and hence the bus scan) as I got my new Neotron Pico expansion cards today.
